File: app/src/main/java/com/dante/diary/net/TimeApi.java
package com.dante.diary.net;

import com.dante.diary.model.Comment;
import com.dante.diary.model.Diary;
import com.dante.diary.model.Notebook;
import com.dante.diary.model.TipResult;
import com.dante.diary.model.User;

import java.util.List;
import java.util.Map;

import androidx.annotation.Nullable;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Response;
import retrofit2.http.DELETE;
import retrofit2.http.Field;
import retrofit2.http.FieldMap;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.Multipart;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Part;
import retrofit2.http.Path;
import retrofit2.http.Query;
import rx.Observable;

/**
 * API for time pill.
 */

public interface TimeApi {

    @GET("topic")
    Observable<Response<ResponseBody>> getTopic();

    @GET("topic/diaries")
    Observable<DiariesResult<List<Diary>>> getTopicDiaries(@Query("page") int page, @Query("page_size") int pageSize);

    @GET("diaries/today")
    Observable<DiariesResult<List<Diary>>> allTodayDiaries(@Query("page") int page, @Query("page_size") int pageSize);

    @GET("diaries/search")
    Observable<NotebooksResult<List<Diary>>> search(@Query("keywords") String keywords, @Query("page") int pageNum,
                                                    @Query("page_size") int pageSize, @Query("notebook_id") Integer notebookId);

    @GET("users/{user_id}/diaries")
    Observable<List<Diary>> getTodayDiaries(@Path("user_id") int userId);

    @GET("diaries/follow")
    Observable<DiariesResult<List<Diary>>> getFollowingDiaries(@Query("page") int page, @Query("page_size") Integer pageSize);

    @GET("relation")
    Observable<UsersResult<List<User>>> getFollowings(@Query("page") int page, @Query("page_size") int pageSize);

    @GET("relation/reverse")
    Observable<UsersResult<List<User>>> getMyFollowers(@Query("page") int page, @Query("page_size") int pageSize);

    @POST("relation/{user_id}")
    Observable<Response<ResponseBody>> follow(@Path("user_id") int userId);

    @FormUrlEncoded
    @POST("reports")
    Observable<Response<ResponseBody>> reportDiary(@Field("user_id") int userId, @Field("diary_id") int diaryId);

    @GET("relation/{user_id}")
    Observable<Response<ResponseBody>> hasfollow(@Path("user_id") int userId);

    @DELETE("relation/{user_id}")
    Observable<Response<ResponseBody>> unfollow(@Path("user_id") int userId);

    @GET("diaries/{diary_id}")
    Observable<Diary> getDiaryDetail(@Path("diary_id") int diaryId);

    @GET("diaries/{diary_id}/comments")
    Observable<List<Comment>> getDiaryComments(@Path("diary_id") int diaryId);

    @FormUrlEncoded
    @POST("diaries/{diary_id}/comments")
    Observable<Comment> postDiaryComment(@Path("diary_id") int diaryId, @FieldMap Map<String, Object> data);

    @DELETE("comments/{id}")
    Observable<Response<ResponseBody>> deleteComment(@Path("id") int commentId);

    @GET("users/my")
    Observable<User> getMyProfile();

    @GET("users/{user_id}")
    Observable<User> getProfile(@Path("user_id") int userId);

    @GET("users/{user_id}/notebooks")
    Observable<List<Notebook>> getMyNotebooks(@Path("user_id") int userId);

    @GET("notebooks/{notebook_id}/diaries")
    Observable<NotebooksResult<List<Diary>>> getDiariesOfNotebook(@Path("notebook_id") int notebookId, @Query("page") int page, @Query("page_size") int pageSize);

    @FormUrlEncoded
    @PUT("notebooks/{id}")
    Observable<Notebook> updateNotebook(@Path("id") int notebookId, @FieldMap Map<String, Object> data);

    @FormUrlEncoded
    @POST("notebooks")
    Observable<Notebook> createNotebook(@FieldMap Map<String, Object> data);

    @DELETE("notebooks/{id}")
    Observable<Response<ResponseBody>> deleteNotebook(@Path("id") int notebookId);

    @Multipart
    @POST("notebooks/{id}/cover")
    Observable<Notebook> setNotebookCover(@Path("id") int notebookId, @Part MultipartBody.Part file);

    @Multipart
    @POST("notebooks/{book_id}/diaries")
    Observable<Diary> createDiary(@Path("book_id") int notebookId, @Part("content") RequestBody content, @Part("join_topic") RequestBody topicState,
                                  @Nullable @Part MultipartBody.Part file);


    @DELETE("diaries/{id}")
    Observable<Response<ResponseBody>> deleteDiary(@Path("id") int diaryId);

    @FormUrlEncoded
    @PUT("diaries/{id}")
    Observable<Diary> updateDiary(@Path("id") int diaryId, @Field("content") String content, @Field("notebook_id") int notebookId);

    @GET("tip")
    Observable<List<TipResult>> getTips();

    @GET("tip/history")
    Observable<List<TipResult>> getTipsHistory();

    @POST("tip/read/{ids}")
    Observable<Response<ResponseBody>> tipsRead(@Path("ids") String ids);

    @FormUrlEncoded
    @POST("users")
    Observable<User> register(@Field("email") String email, @Field("name") String nickName, @Field("password") String password);

    @FormUrlEncoded
    @PUT("users")
    Observable<User> updateUserInfo(@Nullable @Field("name") String name, @Field("intro") String intro);

    @Multipart
    @POST("users/icon")
    Observable<User> setUserIcon(@Part MultipartBody.Part file);

    @DELETE("relation/reverse/{id}")
    Observable<Response<ResponseBody>> cancelFollowed(@Path("id") int id);

    class DiariesResult<T> {
        public int count;
        public int page;
        public int page_size;
        public T diaries;
    }

    class NotebooksResult<T> {
        public int count;
        public int page;
        public int page_size;
        public T items;
    }

    class UsersResult<T> {
        public int count;
        public T users;

    }


}

File: app/src/main/java/com/example/myapplication/net/TimeApi.kt
package com.example.myapplication.net

import com.example.myapplication.data.model.*
import okhttp3.MultipartBody
import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.Response
import retrofit2.http.*

/**
 * API for time pill.
 */
interface TimeApi {
    @GET("topic")
    suspend fun getTopic(): Topic

    @GET("topic/diaries")
    suspend fun getTopicDiaries(
        @Query("page") page: Int,
        @Query("page_size") pageSize: Int
    ): DiariesResult<List<Diary>>

    @GET("diaries/today")
    suspend fun allTodayDiaries(
        @Query("page") page: Int,
        @Query("page_size") pageSize: Int
    ): DiariesResult<List<Diary>>

    @GET("diaries/search")
    suspend fun search(
        @Query("keywords") keywords: String, @Query("page") pageNum: Int,
        @Query("page_size") pageSize: Int, @Query("notebook_id") notebookId: Int?
    ): NotebooksResult<List<Diary>>

    @GET("users/{user_id}/diaries")
    suspend fun getTodayDiaries(@Path("user_id") userId: Int): List<Diary>

    @GET("diaries/follow")
    suspend fun getFollowingDiaries(
        @Query("page") page: Int,
        @Query("page_size") pageSize: Int?
    ): DiariesResult<List<Diary>>

    @GET("relation")
    suspend fun getFollowings(
        @Query("page") page: Int,
        @Query("page_size") pageSize: Int
    ): UsersResult<List<User>>

    @GET("relation/reverse")
    suspend fun getMyFollowers(
        @Query("page") page: Int,
        @Query("page_size") pageSize: Int
    ): UsersResult<List<User>>

    @POST("relation/{user_id}")
    suspend fun follow(@Path("user_id") userId: Int): Response<ResponseBody>

    @FormUrlEncoded
    @POST("reports")
    suspend fun reportDiary(
        @Field("user_id") userId: Int,
        @Field("diary_id") diaryId: Int
    ): Response<ResponseBody>

    @GET("relation/{user_id}")
    suspend fun hasfollow(@Path("user_id") userId: Int): Response<ResponseBody>

    @DELETE("relation/{user_id}")
    suspend fun unfollow(@Path("user_id") userId: Int): Response<ResponseBody>

    @GET("diaries/{diary_id}")
    suspend fun getDiaryDetail(@Path("diary_id") diaryId: Int): Diary

    @GET("diaries/{diary_id}/comments")
    suspend fun getDiaryComments(@Path("diary_id") diaryId: Int): List<Comment>

    @FormUrlEncoded
    @JvmSuppressWildcards
    @POST("diaries/{diary_id}/comments")
    suspend fun postDiaryComment(
        @Path("diary_id") diaryId: Int,
        @FieldMap data: Map<String, Any>
    ): Comment

    @DELETE("comments/{id}")
    suspend fun deleteComment(@Path("id") commentId: Int): Response<ResponseBody>

    @GET("users/my")
    suspend fun getMyProfile(): User

    @GET("users/{user_id}")
    suspend fun getProfile(@Path("user_id") userId: Int): User

    @GET("users/{user_id}/notebooks")
    suspend fun getMyNotebooks(@Path("user_id") userId: Int): List<Notebook>

    @GET("notebooks/{notebook_id}/diaries")
    suspend fun getDiariesOfNotebook(
        @Path("notebook_id") notebookId: Int,
        @Query("page") page: Int,
        @Query("page_size") pageSize: Int
    ): NotebooksResult<List<Diary>>

    @FormUrlEncoded
    @JvmSuppressWildcards
    @PUT("notebooks/{id}")
    suspend fun updateNotebook(
        @Path("id") notebookId: Int,
        @FieldMap data: Map<String, Any>
    ): Notebook

    @FormUrlEncoded
    @JvmSuppressWildcards
    @POST("notebooks")
    suspend fun createNotebook(@FieldMap data: Map<String, Any>): Notebook

    @DELETE("notebooks/{id}")
    suspend fun deleteNotebook(@Path("id") notebookId: Int): Response<ResponseBody>

    @Multipart
    @POST("notebooks/{id}/cover")
    suspend fun setNotebookCover(
        @Path("id") notebookId: Int,
        @Part file: MultipartBody.Part
    ): Notebook

    @Multipart
    @POST("notebooks/{book_id}/diaries")
    suspend fun createDiary(
        @Path("book_id") notebookId: Int,
        @Part("content") content: RequestBody?,
        @Part("join_topic") topicState: RequestBody?,
        @Part file: MultipartBody.Part?
    ): Diary

    @DELETE("diaries/{id}")
    suspend fun deleteDiary(@Path("id") diaryId: Int): Response<ResponseBody>

    @FormUrlEncoded
    @PUT("diaries/{id}")
    suspend fun updateDiary(
        @Path("id") diaryId: Int,
        @Field("content") content: String,
        @Field("notebook_id") notebookId: Int
    ): Diary

    @GET("tip")
    suspend fun getTips(): List<TipResult>

    @GET("tip/history")
    suspend fun getTipsHistory(): List<TipResult>

    @POST("tip/read/{ids}")
    suspend fun tipsRead(@Path("ids") ids: String): Response<ResponseBody>

    @FormUrlEncoded
    @POST("users")
    suspend fun register(
        @Field("email") email: String,
        @Field("password") password: String,
        @Field("name") nickName: String
    ): Response<User>

    @FormUrlEncoded
    @PUT("users")
    suspend fun updateUserInfo(@Field("name") name: String, @Field("intro") intro: String): User

    @Multipart
    @POST("users/icon")
    suspend fun setUserIcon(@Part file: Part?): User

    @DELETE("relation/reverse/{id}")
    suspend fun cancelFollowed(@Path("id") id: Int): Response<ResponseBody>

}